File: csharp/exercise/exer04-dining-philosophers.cs
/*
 * THE DINING PHILOSOPHERS PROBLEM
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Exer04 : IRunnable
{
    public void run()
    {
        const int NUM_PHILOSOPHERS = 5;


        var chopstick = new Semaphore[NUM_PHILOSOPHERS];

        for (int i = 0; i < NUM_PHILOSOPHERS; ++i)
            chopstick[i] = new Semaphore(1, NUM_PHILOSOPHERS + 1);


        var lstTh = new List<Thread>();

        for (int ith = 0; ith < NUM_PHILOSOPHERS; ++ith)
        {
            int i = ith;

            lstTh.Add(new Thread(() =>
            {
                int n = NUM_PHILOSOPHERS;
                Thread.Sleep(1000);

                chopstick[i].WaitOne();
                chopstick[(i + 1) % n].WaitOne();

                Console.WriteLine($"Philosopher #{i} is eating the rice");

                chopstick[(i + 1) % n].Release();
                chopstick[i].Release();
            }));
        }


        lstTh.ForEach(th => th.Start());
    }
}

File: cpp/cpp-std/exer04-dining-philosophers.cpp
/*
THE DINING PHILOSOPHERS PROBLEM
*/


#include <iostream>
#include <chrono>
#include <thread>
#include <mutex>
using namespace std;



void doTaskPhilosopher(std::mutex chopstick[], int numPhilo, int idPhilo) {
    int n = numPhilo;
    int i = idPhilo;

    std::this_thread::sleep_for(std::chrono::seconds(1));

    chopstick[i].lock();
    chopstick[(i + 1) % n].lock();

    cout << "Philosopher #" << i << " is eating the rice" << endl;

    chopstick[i].unlock();
    chopstick[(i + 1) % n].unlock();
}



int main() {
    constexpr int NUM_PHILOSOPHERS = 5;

    std::mutex chopstick[NUM_PHILOSOPHERS];
    std::thread lstTh[NUM_PHILOSOPHERS];

    // CREATE THREADS
    for (int i = 0; i < NUM_PHILOSOPHERS; ++i) {
        lstTh[i] = std::thread(&doTaskPhilosopher, chopstick, NUM_PHILOSOPHERS, i);
    }

    // JOIN THREADS
    for (auto&& th : lstTh) {
        th.join();
    }

    return 0;
}
